app/src/main/java/io/homeassistant/companion/android/sensors/SensorsSettingsFragment.kt

```diff
@@ -56,6 +103,37 @@ class SensorsSettingsFragment : PreferenceFragmentCompat() {
 
         setPreferencesFromResource(R.xml.sensors, rootKey)
 
+        findPreference<SwitchPreference>("enable_disable_sensors")?.let {
+
+            var permArray: Array<String> = arrayOf()
+            it.setOnPreferenceChangeListener { _, newState ->
+                val enabledAll = newState as Boolean
+                val sensorDao = AppDatabase.getInstance(requireContext()).sensorDao()
+
+                SensorReceiver.MANAGERS.forEach { managers ->
+                    managers.availableSensors.forEach { basicSensor ->
+                        var sensorEntity = sensorDao.get(basicSensor.id)
+
+                        if (!managers.checkPermission(requireContext(), basicSensor.id))
+                            permArray += managers.requiredPermissions(basicSensor.id).asList()
+
+                        if (sensorEntity != null) {
+                            sensorEntity.enabled = enabledAll
+                            sensorEntity.lastSentState = ""
+                            sensorDao.update(sensorEntity)
+                        } else {
+                            sensorEntity = Sensor(basicSensor.id, enabledAll, false, "")
+                            sensorDao.add(sensorEntity)
+                        }
+                    }
+                }
+                if (!permArray.isNullOrEmpty())
+                    requestPermissions(permArray, 0)
```

Comment:
This should probably have have something similar to the detail screen that listens for the permission response and toggles off the preference if not all permissions were granted rather than waiting for the update method to be called again..

Reply:
So originally I had taken this from the details screen but due to the for loop I had to take the permission request out and make them after we are done gathering all the required permissions.  I can look at the `onRequestPermissionsResult` to see if it makes the update faster there.